File: python/test/fdbus_test_client.py
"""
/*
 * Copyright (C) 2015   Jeremy Chen jeremy_cz@yahoo.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"""

import fdbus
import sys
import time
import os
import Example_pb2 as ex

def process_response(code, resp_data, tag):
    if code == ex.REQ_METADATA:
        resp_info = ex.NowPlayingDetails()
        resp_info.ParseFromString(resp_data)
        print(tag, resp_info)
    else:
        print('unknown message code is received: ', code)

class MyTestClient(fdbus.FdbusClient):
    def __init__(self, name):
        super(MyTestClient, self).__init__(name)
        
    def onOnline(self, sid):
        print('MyTestClient: onOnline')
        subscribe_list = [{'event_code' : ex.NTF_ELAPSE_TIME,         'topic' : None},
                          {'event_code' : ex.NTF_MEDIAPLAYER_CREATED, 'topic' : 'topic-1'},
                          {'event_code' : ex.NTF_MEDIAPLAYER_CREATED, 'topic' : 'topic-2'}]
        self.subscribe(subscribe_list)
        super(MyTestClient, self).onOnline(sid)

    def onOffline(self, sid):
        print('MyTestClient: onOffline')
        super(MyTestClient, self).onOffline(sid)

    def onReply(self, sid, msg_code, msg_data, status, user_data):
        if status == 0:
            process_response(msg_code, msg_data, 'onResply: ')
        
    def onBroadcast(self, sid, event_code, event_data, topic):
        if event_code == ex.NTF_ELAPSE_TIME:
            et = ex.ElapseTime()
            et.ParseFromString(event_data)
            print('onBroadcast: \n', et)
        elif event_code == ex.NTF_MEDIAPLAYER_CREATED:
            print('onBroadcast: event: ', event_code, ', topic: ', topic)
        else:
            print('onBroadcast - unknown event: ', event_code)
        
fdbus.fdbusStart(os.getenv('FDB_CLIB_PATH'))
client_list = []
nr_clients = len(sys.argv) - 1
for i in range(nr_clients):
    name = sys.argv[i+1]
    url = 'svc://' + name
    client = MyTestClient(name)
    client_list.append(client)
    client.connect(url)
    
song_id = 0
while True:
    req = ex.SongId()
    for i in range(nr_clients):
        req.id = song_id
        song_id += 1
        #client_list[i].invoke_async(200, 'hello, world')
        client_list[i].invoke_async(ex.REQ_METADATA, req.SerializeToString())
        time.sleep(1)

        req.id = song_id
        song_id += 1
        msg = client_list[i].invoke_sync(ex.REQ_METADATA, req.SerializeToString())
        if msg['status']:
            print('Error for sync invoke: ', msg['status'])
        else:
            process_response(msg['msg_code'], msg['msg_data'], 'sync reply: ')
     
        fdbus.releaseReturnMsg(msg)
        fdbus.FDB_LOG_E('fdb_py_clt', 'name: ', 'sdcard', 'size', 1000)
        time.sleep(1)

File: c/test/fdbus_test_client.c
/*
 * Copyright (C) 2015   Jeremy Chen jeremy_cz@yahoo.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdlib.h>
#include <stdio.h>
#define FDB_LOG_TAG "FDB-CCLT"
#include <common_base/fdbus_clib.h>
#include "fdbus_test_msg_ids.h"

static void on_online(fdb_client_t *self, FdbSessionId_t sid)
{
    fdb_subscribe_item_t items[] = {{FDB_TEST_EVENT_ID_1, "topic1"},
                                    {FDB_TEST_EVENT_ID_2, "topic2"},
                                    {FDB_TEST_EVENT_ID_3, "topic3"}};
    fdb_client_subscribe(self, items, Fdb_Num_Elems(items));
    FDB_LOG_D("on online: %d\n", sid);
}

static void on_offline(fdb_client_t *self, FdbSessionId_t sid)
{
    FDB_LOG_D("on offline: %d\n", sid);
}

static void on_reply(fdb_client_t *self,
                     FdbSessionId_t sid,
                     FdbMsgCode_t msg_code,
                     const uint8_t *msg_data,
                     int32_t data_size,
                     int32_t status,
                     void *user_data)
{
    int32_t i;

    FDB_LOG_D("on reply: sid: %d, code: %d, size: %d\n", sid, msg_code, data_size);
    for (i = 0; i < data_size; ++i)
    {
        FDB_LOG_D("    data received: %d\n", msg_data[i]);
    }
}

static void on_broadcast(fdb_client_t *self,
                         FdbSessionId_t sid,
                         FdbMsgCode_t msg_code,
                         const uint8_t *msg_data,
                         int32_t data_size,
                         const char *filter)
{
    int32_t i;

    FDB_LOG_D("on broadcast: sid: %d, code: %d, size: %d, filter: %s\n",
                sid, msg_code, data_size, filter);
    for (i = 0; i < data_size; ++i)
    {
        FDB_LOG_D("    data received: %d\n", msg_data[i]);
    }
}

int main(int argc, char **argv)
{
    int32_t i;

    fdb_start();
    fdb_client_t **client_array = (fdb_client_t **)malloc(sizeof(fdb_client_t *) * argc);
    for (i = 0; i < (argc - 1); ++i)
    {
        char url[1024];
        snprintf(url, sizeof(url), "svc://%s", argv[i + 1]);
        client_array[i] = fdb_client_create(argv[i + 1], 0);
        fdb_client_register_event_handle(client_array[i], on_online, on_offline, on_reply, 0, on_broadcast);
        fdb_client_connect(client_array[i], url);
    }

    uint8_t buffer[19];
    uint8_t count = 0;
    uint32_t msg_code = 0;
    while (1)
    {
        int32_t i;

        for (i = 0; i < Fdb_Num_Elems(buffer); ++i)
        {
            buffer[i] = count++;
        }
        for (i = 0; i < (argc - 1); ++i)
        {
            fdb_client_invoke_async(client_array[i], msg_code++, buffer, Fdb_Num_Elems(buffer), 0, 0, 0);
        }
        sysdep_sleep(111);
    }
}
